{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Project 2 - Genetic Algorithm for Image Reconstruction.**\n",
    "**Subject:** MC906/MO416 - Introduction to Artificial Intelligence \n",
    "\n",
    "**Authors:**\n",
    "\n",
    "    Eduardo Barros Innarelli - RA 170161\n",
    "    Victor Ferreira Ferrari  - RA 187890\n",
    "    Vinicius Couto Espindola - RA 188115"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Introduction**\n",
    "### Project\n",
    "The project consists of a study of genetic algorithm-based solutions for **image reconstruction**. A few set aspects of the problem will be explained, implemented and discussed here, such as:\n",
    "\n",
    "- The **modeling** of the problem as an evolutionary problem (chromosomes, genes, etc);\n",
    "- How to **generate the initial population**;\n",
    "- The chosen **fitness function**;\n",
    "- How to **visualize** the result.\n",
    "\n",
    "Along with that, some aspects of the genetic algorithm will be tested and discussed with multiple approaches, such as:\n",
    "\n",
    "- Stop criteria;\n",
    "- Selection technique;\n",
    "- Crossover technique;\n",
    "- Mutation technique;\n",
    "- Generation replacement method.\n",
    "\n",
    "Also, the following **parameters** will be tested with multiple values as well:\n",
    "- Population size;\n",
    "- Mutation rate;\n",
    "- Crossover rate.\n",
    "\n",
    "In the end, different variations of the process will be compared and discussed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage\n",
    "\n",
    "This project uses the external libraries _OpenCV_ for image reading and manipulation, _NumPy_ for array manipulation, and _MatPlotLib_ for result visualization. The other imports are part of the Standard Python Library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# External Libraries\n",
    "from matplotlib import pyplot as plt\n",
    "from cv2 import imread, imwrite, resize\n",
    "import numpy as np\n",
    "\n",
    "# Python Libraries\n",
    "from time import time\n",
    "from itertools import combinations\n",
    "from random import sample, choices\n",
    "from os.path import join, basename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Problem**\n",
    "### General Information\n",
    "The problem consists of, given an image, recreate it from scratch, using the original image only for fitness purposes, to check the similarity between our image and the original. This is not particularly useful in many applications, but can be used as basis for other problems, and is good for the purpose of this project: test different methods of reproduction, mutation, etc.\n",
    "\n",
    "The motivation came from [Roger Johansson's \"Evolution of Mona Lisa\"](https://rogerjohansson.blog/2008/12/07/genetic-programming-evolution-of-mona-lisa/), but instead of using polygons, the pixel values are directly used.\n",
    "\n",
    "### Modeling\n",
    "The modeling of the problem as a GA problem is almost direct: the image is the individual (chromosome) and each of its pixels is a gene. The image can be represented by a multitude of different ways. The chosen way of representation is as a flat array, as this allows for easier operations, especially crossover.\n",
    "\n",
    "### Inputs\n",
    "Due to the huge number of combinations in an image (255 possiblities per pixel), greyscale images are used in this project. Low-resolution images are best for this, but there are methods of reducing the resolution using _OpenCV_."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Methods**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes on Structure\n",
    "\n",
    "The code was organized in such a manner which allows us to first define every method used as well as discuss them in separated modules, making each part more readable and comprehesible.\n",
    "\n",
    "In order to merge all components of the code we created a class called GeneticAlgorithm. This class retains some essential variables for the methods defined along the report. To allow each method to access these variables, even though they are not a part of the class, nearly every method must receive as it's first parameter the class instance which is referred as the parameter 'self' in their definitions. It should also be noted that the method usualy have no return statements, simply because the 'self' parameter is a reference to the class instance, which allow the methods to make changes in-place.\n",
    "\n",
    "To specify which class variables are necessary and which are defined by the methods in each section, there will be a list of **used class variables** and **defined class variables** within the description of every section. On top of these uses and definitions, there are parameters which are method dependent, these are in the sections **necessary parameters**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population Initialization\n",
    "\n",
    "The first step to run a genetic algorithm is creating a population to serve as a starting point for the selection, reproduction and mutation process. This function also handles the initilization of the current best population, the best population overall fitness and best individual fitness, which are paramount for convergence check and final results.   \n",
    "\n",
    "**Necessary parameters:**\n",
    " 1. *pop_size* - Defines the amount of individuals to be created in the initial population.\n",
    "\n",
    "**Used class variables:** \n",
    " - None.\n",
    "\n",
    "**Defined class variables:**\n",
    " 1. *self.population* - List of *ndarrays* where each element is an individual.\n",
    " 2. *self.all_fits* - List of float values where each element is the fittnes of an individual.\n",
    " 3. *self.pop_size* - Holds the value chosen for population size.\n",
    " \n",
    "**Note:** The amount of genes in each individual is defined by the goal variables in the class. These variables store information of the image we aim to generate, including the amount of pixels (which are the genes in our model).\n",
    "\n",
    "The following code defines some methods which will be used to create a population out of nothing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random(self, pop_size):\n",
    "    # Generate completely random indivuduals \n",
    "    self.population = []\n",
    "    size = self.goal.size\n",
    "    \n",
    "    for i in range(pop_size):\n",
    "        print(f\"Generating Population...  {i:>3}/{pop_size}\\r\", end='')\n",
    "        person = np.random.randint(0, 255, size, dtype=np.uint8)\n",
    "        self.population.append(person)\n",
    "    self.pop_size = pop_size\n",
    "    print(f\"{' '*50}\\r\", end='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def white(self, pop_size):\n",
    "    self.population = []\n",
    "    size = self.goal.size\n",
    "\n",
    "    for i in range(pop_size):\n",
    "        print(f\"Generating Population...  {i:>3}/{pop_size}\\r\", end='')\n",
    "        person = np.full(size, 255, dtype=np.uint8)\n",
    "        self.population.append(person)\n",
    "    self.pop_size = pop_size\n",
    "    print(f\"{' '*50}\\r\", end='')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fitness Function\n",
    "**Responsible: Victor**\n",
    "\n",
    "For each individual, we must determine a method which will tell us how apt such individual for the problem in question.\n",
    "\n",
    "The following methods define some method to evaluate how close is an individual to the image we seek to generate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def squared_error(individual, original):\n",
    "     return ((original.astype(np.float64) - individual.astype(np.float64))**2).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mse(individual, original):\n",
    "     return squared_error(individual,original)/individual.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_pixels(individual, original):\n",
    "    return np.count_nonzero(original.astype(np.int16) - individual.astype(np.int16))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convergence Criteria\n",
    "\n",
    "The algorithm cannot run forever. In this section will define a convergence method for halting the program's exection. \n",
    "\n",
    "The convergence function must identify when the algorithm's improvement rate has significantly decreased or even stalled. To do so, it must know to some extent the rate of improvement per generation, which can be infered by the difficulty in improving the best individual found so far. Assume we have the best indivudual so far ($B$), therefore We can estimate when the algorithm is decreasing it's evolution rate (or improvement rate) by checking how many generations have passed since a better individual than $B$ was achieved: the more generations where $B$ wasnt improved, the more likely is has converged. The convergence method defined in this section is based on this idea.\n",
    "\n",
    "**Necessary parameters:**\n",
    " 1. *limit* - Suppose generation $X$ has the best individual so far, now assume we are at generation $Y$ and $Y>X$. This implies that the last $L=Y-X$ generation did not improve the fittest individual found so far. If $L>limit$ we consider that the algorithm has converged.\n",
    "\n",
    "**Used class variables:** \n",
    " 1. *self.all_fits* - Used to find the fittest individual of the new generation.\n",
    " 2. *self.best_fit* - Holds the fittness of the best individual across all generations.\n",
    "\n",
    "**Defined class variables:**\n",
    " 1. *self.counter* - Counts generations since the last improvement of the best individual.\n",
    " 2. *self.best_genfit* - If theres a new better individual we must update this variable.\n",
    "\n",
    "**Note:** even checking for convergence, the algorithm might still present sufficiently small improvements to avoid convergence, however, they might be so small that it will take a enormous amount of time to converge. As a failsafe to this scenario, we will limit the total amount of generations that can be created from withing the GeneticAlgorithm class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convergence(self, limit):\n",
    "    ''' Identifies convergence and updates fittest individual found so far \n",
    "        \n",
    "        Parameters:\n",
    "            self - GeneticAlgorithm class which the function is being called from\n",
    "            limit - Amount of iterations wich can be worst than the best one found so far\n",
    "        Return:\n",
    "            Bool - True if converged, False otherwise.\n",
    "    '''\n",
    "    # Get min fitness for population\n",
    "    new_genfit = self.all_fits.min()\n",
    "     \n",
    "    # Update number of iterations without improvement\n",
    "    if new_genfit >= self.best_genfit:\n",
    "        self.counter += 1\n",
    "    else:\n",
    "        self.counter = 0\n",
    "        self.best_genfit = new_genfit\n",
    "    \n",
    "    # Check convergence\n",
    "    if self.counter >= limit or self.best_genfit <= 0:\n",
    "        self.counter = 0\n",
    "        return True\n",
    "    else:\n",
    "        return False "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parents Selection\n",
    "\n",
    "In this section we'll define methods responsible for selecting individual of the population responsible for bearing the next generation. The main goal it to find a method which properly balances the diversity and quality of individuals.\n",
    "\n",
    "A premature analisys of the fitness values presented by each individual in the population showed that the fitness value for the individuals tend to be very close in the model we created, this can affect methods such as *roulette selection* which depend heavily on the difference between the fitness score of individuals. In order to prevent this issue, we decided to implement methods which depend only on the comparison among individuals: the values might be very similar, but as long as we cand define which one is smaller or larger, we can use the method realiably.\n",
    "\n",
    "**Necessary parameters:**\n",
    " 1. **P:** Defines the probability of accepting a certain idividual.\n",
    " 2. **N:** Amount of couples to be selected for mating.\n",
    " 3. **K:** Amount of individuals to partake in a competition.\n",
    " \n",
    "**Used class variables:** \n",
    " 1. *self.pop_size:* will select individuals to mate and reference then by index.\n",
    " 2. *self.all_fits:* the selection is based on the individuals fitness.\n",
    " \n",
    "**Defined class variables:**\n",
    " 1. *self.mates:* creates a list of tuples where each tuple pairs the index of two individual for reproduction.\n",
    " \n",
    "The following code defines some methods which will be used to select individual for the creation of the next generation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if the tournament size is larger, weak individuals have \n",
    "# a smaller chance to be selected, because, if a weak \n",
    "# individual is selected to be in a tournament, there is \n",
    "# a higher probability that a stronger individual is also \n",
    "# in that tournament.\n",
    "\n",
    "def tournament(self, k, N, P, T):\n",
    "    pool = list(range(self.pop_size))\n",
    "    pairs = {x:set() for x in pool}\n",
    "    fitness = lambda x: self.all_fits[x]\n",
    "    winners,self.mates = [],[]\n",
    "\n",
    "    # Run tournament to pick N winners\n",
    "    while len(winners) < N:\n",
    "        competitors = np.random.choice(pool, size=k, replace=False)\n",
    "        winner = max(competitors, key=fitness)\n",
    "        winners.append(winner)\n",
    "        pool.remove(winner) # remove winner from pool\n",
    "    \n",
    "    # Combination: given a pool of winners, match them as couples\n",
    "    #  - Each individual can match with at most P other individuals\n",
    "    #  - Each individual cannot match twice with some individual\n",
    "    #  - Each individual cannot match with himself\n",
    "    # There can be at most T matches, but it might be less due to contraints \n",
    "    for (dad,mom) in combinations(winners, 2):\n",
    "        if (len(pairs[dad]) < P) and (len(pairs[mom]) < P): \n",
    "            pairs[dad].add(mom)\n",
    "            pairs[mom].add(dad)\n",
    "            self.mates.append((dad,mom))\n",
    "            if len(self.mates) == T: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ranked(self):\n",
    "    fit = self.all_fits\n",
    "    size = self.pop_size\n",
    "    \n",
    "    # Roulette method: select N/2 individuals from the population,\n",
    "    # with proportional probabilities to their fitness values.\n",
    "    probs = fit/fit.sum()\n",
    "    probs = (1-probs)/(size-1)\n",
    "    winners = np.random.choice(np.arange(size), size=size//2, replace=True, p=probs)    \n",
    "    \n",
    "    self.mates = list(zip(winners[::2],winners[1::2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Crossover Techniques\n",
    "\n",
    "In this section we'll define methods that generate descendants from selected parents. \n",
    "\n",
    "Based on genetic recombination, crossover exchanges genes between two individuals in a stochastic manner, resulting in a new one. Computationally, the way genes are swapped depends on the method implemented. By reproducting, the population preserves genetic diversity between generations.\n",
    "\n",
    "**Necessary parameters:**\n",
    " 1. *K* - Determines the ammount of points to use when applying k-point crossover.\n",
    " 2. *exchange_prob* - Probability of exchanging a gene when applying uniform crossover.\n",
    " \n",
    "**Used class variables:** \n",
    " 1. *self.goal.size* - This attribute tells us how many genes are there in a individual.\n",
    " 1. *self.mates* - Each couple in this variable will be parents in the reproduction method.\n",
    " \n",
    "**Defined class variables:**\n",
    " 1. *self.children* - These are the new individuals generated from reproduction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def k_point_crossover(self, k):\n",
    "    ''' \n",
    "    For each couple, K points are picked randomly from their chromossomes and\n",
    "    genes between these points are swapped, intercalating pairs of points.\n",
    "    \n",
    "    Parameters:\n",
    "        self - GeneticAlgorithm class which the function is being called from\n",
    "        k - Amount of crossover points for gene exchange\n",
    "    '''\n",
    "    \n",
    "    num_genes = self.goal.size\n",
    "    self.children = []\n",
    "\n",
    "    for father, mother in self.mates:\n",
    "        # Save k unique random crosspoints in a sorted list\n",
    "        crosspoints = sorted(sample(range(0, num_genes-1), k))\n",
    "        \n",
    "        # Deep copy to freely change values\n",
    "        father = self.population[father].copy()\n",
    "        mother = self.population[mother].copy()\n",
    "        \n",
    "        # List of crosspoint pairs, adding a pair with last crosspoint\n",
    "        # and last pixel\n",
    "        # Ex: \n",
    "        # 16 pixels, points [1,9,12] -> pairs [(1,9), (9,12), (12,16)]\n",
    "        pairs = list(zip(crosspoints,crosspoints[1:]))\n",
    "        pairs.append(tuple((crosspoints[-1], num_genes)))\n",
    "\n",
    "        # Perform crossover in pair range, intercalating pairs\n",
    "        for start, end in pairs[::2]:\n",
    "            father[start:end], mother[start:end] = mother[start:end], father[start:end]\n",
    "        \n",
    "        self.children += [father, mother]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniform_crossover(self, exchange_prob):\n",
    "    '''\n",
    "    For each couple, it's genes are exchanged with a given probability.\n",
    "    \n",
    "    Parameters:\n",
    "        self - GeneticAlgorithm class which the function is being called from\n",
    "        exchange_prob - Probability of gene being exchanged between parents\n",
    "    '''\n",
    "    \n",
    "    num_genes = self.goal.size\n",
    "    self.children = []\n",
    "    \n",
    "    for father, mother in self.mates:\n",
    "        \n",
    "        # Deep copy to freely change values\n",
    "        father = self.population[father].copy()\n",
    "        mother = self.population[mother].copy()\n",
    "        \n",
    "        # Exchange genes with a given probability\n",
    "        for gene in range(num_genes):\n",
    "            if np.random.uniform() < exchange_prob:\n",
    "                father[gene], mother[gene] = mother[gene], father[gene]\n",
    "                \n",
    "        self.children += [father, mother]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation Method\n",
    "**Responsible: Victor**\n",
    "\n",
    "To simply permute the genes of the previous generation might reduce the diversity among individuals by creating childs which are too similar to their parents. To counter this we'll use mutation methods allowing genes to mix within individuals. \n",
    "\n",
    "**Necessary parameters:**\n",
    " 1. *MR* - Mutation rate dictates the probability of an individual to mutate.\n",
    " 2. *MA* - Mutation amount defines the amount of genes which will mutate.\n",
    " \n",
    "**Used class variables:** \n",
    " 1. *self.goal.size* - This attribute tells us how many genes are there in a individual.\n",
    " 1. *self.children* - Mutations are applied only on the individuals resulting from the reproduction step.\n",
    " \n",
    "**Defined class variables:**\n",
    " 1. *self.children* - The childrens gene are altered in-place altering this value.\n",
    "\n",
    "The following code defines some methods which will be used to mutate the individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_mut(self, MR, MA):\n",
    "    # Set some synonyms\n",
    "    size = self.goal.size\n",
    "    pop  = self.children\n",
    "    \n",
    "    # Select individuals to mutate.\n",
    "    mut_amount = int(MR*len(pop)) # amount of individuals to mutate\n",
    "    individuals = choices(pop, k=mut_amount)\n",
    "    \n",
    "    # Change elements to random value.\n",
    "    gene_amount = int(MA*size) # amount of genes to mutate per individual\n",
    "    for i in individuals:\n",
    "        genes = np.random.choice(size, gene_amount, replace=False)\n",
    "        new   = np.random.randint(0, 255, gene_amount, np.uint8)\n",
    "        i[genes] = new # Changes are in-place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reduced_mut(self, MR, MA):\n",
    "    # Set some synonyms\n",
    "    size = self.goal.size\n",
    "    pop  = self.children\n",
    "    \n",
    "    # Select individuals to mutate.\n",
    "    mut_amount = int(MR*len(pop)) # amount of individuals to mutate\n",
    "    individuals = choices(pop, k=mut_amount)\n",
    "    \n",
    "    # Change elements to random value.\n",
    "    gene_amount = int(MA*size) # amount of genes to mutate per individual\n",
    "    for i in individuals:\n",
    "        genes = np.random.choice(size, gene_amount, replace=False)\n",
    "        new   = np.random.randint(0, 31, gene_amount, np.uint8)\n",
    "        i[genes] = new*8 # Changes are in-place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def inc_mut(self, MR, MA):\n",
    "    # Set some synonyms\n",
    "    size = self.goal.size\n",
    "    pop  = self.children\n",
    "    \n",
    "    # Select individuals to mutate.\n",
    "    mut_amount = int(MR*len(pop)) # amount of individuals to mutate\n",
    "    individuals = choices(pop, k=mut_amount)\n",
    "    \n",
    "    # Change elements to random value.\n",
    "    gene_amount = int(MA*size) # amount of genes to mutate per individual\n",
    "    for i in individuals:\n",
    "        genes = np.random.choice(size, gene_amount, replace=False)\n",
    "        new   = np.random.randint(-10, 10, gene_amount, np.int8)\n",
    "        i[genes] += new # Changes are in-place"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Replacement Techniques\n",
    "\n",
    "In this stage we're storing both the parents and their children. Increasing the population size can affect the algorithm performance in terms of memory usage and time as well as the convergence rate due to unfit individuals being kept alive. To mitigate this, it's necessary to choose which individuals will remain in the new generation.\n",
    "\n",
    "**Necessary parameters:**\n",
    " - None.\n",
    " \n",
    "**Used class variables:** \n",
    " 1. *self.children* - Calculates the fittnes of every child to apply the replacement method.\n",
    " \n",
    "**Defined class variables:**\n",
    " 1. *self.all_fits* - Will be updated with the fitness of the individuals that survived the replacement.\n",
    " 2. *self.population* - The population will also be updated with the individuals that survived the replacement. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def elitism(self):\n",
    "    ''' \n",
    "    Preserve the fittest individuals between both parents and children\n",
    "        \n",
    "    Parameters:\n",
    "        self - GeneticAlgorithm class which the function is being called from\n",
    "    '''\n",
    "    \n",
    "    # Calculate children fitness.\n",
    "    generator = map(lambda x: self.fit(x, self.goal), self.children)\n",
    "    all_fits = np.concatenate((self.all_fits, np.array(list(generator), dtype=np.float64)))\n",
    "    \n",
    "    # Sort fits and population by fits.\n",
    "    idxs = all_fits.argsort()\n",
    "    full = self.population + self.children\n",
    "    full = np.array(full)\n",
    "    \n",
    "    # Combine parents with children, sorted by fitness function\n",
    "    combined = full[idxs]\n",
    "    all_fits = all_fits[idxs]\n",
    "    \n",
    "    # Update population with the fittest individuals\n",
    "    self.population = list(combined[:self.pop_size])\n",
    "    self.all_fits = all_fits[:self.pop_size]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def steady_state(self):\n",
    "    ''' \n",
    "    Children replaces the least fit parents.\n",
    "    \n",
    "    Obs: if number of children and parents are equal, this have the same\n",
    "    effect as a generational replacement.\n",
    "    \n",
    "    Parameters:\n",
    "        self - GeneticAlgorithm class which the function is being called from\n",
    "    '''\n",
    "    \n",
    "    pop_sorted = sorted(range(self.pop_size), key = lambda x: self.all_fits[x])\n",
    "    \n",
    "    # Replace worst fits from population with childs\n",
    "    while self.children and pop_sorted:\n",
    "        old = pop_sorted.pop()\n",
    "        new = self.children.pop()\n",
    "        self.population[old] = new\n",
    "        self.all_fits[old] = self.fit(new, self.goal)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Genetic Algorithm Class**\n",
    "\n",
    "The class will be responsible for selecting which methods will be used for each of the methods in each of the algorithms steps, as well as define it's parameters and executing the main loop. The main loop will be responsible for properly calling the methods, any in between tasks as wells as check halting conditions and picking the fittest individual found in all generations.\n",
    "\n",
    "##### Preprocessing and Visualization Method\n",
    "To perceive empirically if the quality of each generation as well as the final result, we will use openCV to exhibit the image represented by some individual.\n",
    "\n",
    "##### Fitness Update Method\n",
    "There are multiple steps in which we must use the fitness of every individual in the population, to avoid overhead, we'll define a function which will take care of updating the fitness values list. Calling it only once or twice (depending on the method for survivors selection) to avoid overhead.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneticAlgorithm():\n",
    "    ''' The GeneticAlgorithm class retains and coordinates all necessary variables and methods\n",
    "        for the execution of the genetic algorithm created.\n",
    "        It's initialization receives the methods which will compute each step of the algorithm\n",
    "        as well as the parameters utilized by each of these methods. A limit integer is also\n",
    "        define to halt the programs execution.\n",
    "        \n",
    "        Methods:\n",
    "            run    - execute the genetic algorithm based on the instance variables/methods\n",
    "            update - recalculate individuals fitness and updates best individual so far\n",
    "            read   - read a image from file, rescale and flatten\n",
    "            show   - reshape flattened image and exhibit with matplotlib\n",
    "    '''\n",
    "    \n",
    "    def __init__(self, \n",
    "                 limit     = 10000, # Max amount of generations to create\n",
    "                 fit       = None,  # Function to calculate individual fitness\n",
    "                 conv      = None,  # Convergence function and parameters\n",
    "                 init      = None,  # Population Initializer function and parameters\n",
    "                 mating    = None,  # Population Selection function and parameters\n",
    "                 crossover = None,  # Selected Indviduals Crossover function and parameters\n",
    "                 mutate    = None,  # Selected Indviduals Mutation function and parameters\n",
    "                 replace   = None,  # Choose which individuals are kept/replaced\n",
    "                ):\n",
    "        \n",
    "        # Halting criterias\n",
    "        self.counter = 0 # Convergence counter\n",
    "        self.limit = limit\n",
    "        self.abort = lambda: bool(self.gen_count >= self.limit)\n",
    "        \n",
    "        # Goal related variables\n",
    "        self.shape = None # Holds original downsized image shape\n",
    "        self.goal  = None # Holds an array with the ideal individual (goal image)\n",
    "        \n",
    "        # Convergence related variables\n",
    "        self.best        = None         # Best individual obtained so far\n",
    "        self.best_fit    = float('inf') # The fit score of the bets individual so far\n",
    "        self.all_fits    = None         # List of each individual fitness\n",
    "        self.best_gen    = None         # Best generation.\n",
    "        self.best_genfit = float('inf')\n",
    "        \n",
    "        # Generation related variables\n",
    "        self.pop_size   = None  # The amount of individuals kept per generation\n",
    "        self.population = None  # List of individual of the current generation\n",
    "        self.gen_count  = 1     # Current generation (first, second, ...)\n",
    "        self.mates      = None  # Paired individuals which will reproduce\n",
    "        self.children   = []\n",
    "    \n",
    "        # Static Genetic Methods Functions\n",
    "        self.fit = fit\n",
    "        define = lambda x: lambda: x[0](self,*x[1]) if x[1] else x[0](self)\n",
    "        self.conv   = define(conv)\n",
    "        self.init   = define(init)\n",
    "        self.mating = define(mating)\n",
    "        self.crossover = define(crossover)\n",
    "        self.mutate = define(mutate)\n",
    "        self.replace = define(replace)\n",
    "        # TODO: MR needs to decrease with time to converge.\n",
    "        \n",
    "        # Timing variables\n",
    "        self.time = 0\n",
    "        \n",
    "        test = [self.conv,self.init,self.mating,self.crossover,self.mutate]\n",
    "        test = all(map(bool, test))\n",
    "        assert test, \"There are undefined steps for the algorithm\"\n",
    "    \n",
    "    def test(self, img_path):\n",
    "        self.time = time()\n",
    "        self.read(img_path, res=0.6)\n",
    "        self.init()\n",
    "        \n",
    "        generator = map(lambda x: self.fit(x, self.goal), self.population)\n",
    "        self.all_fits = np.array(list(generator), dtype=np.float64)\n",
    "        \n",
    "        while not (self.conv() or self.abort()):\n",
    "            print(f\"Current Generation: {self.gen_count:>6} (avg. fit: {self.all_fits.mean()}). Current Best: {self.best_fit:>8.3f} (gen {self.best_gen}).\\r\", end='')\n",
    "            \n",
    "            \n",
    "            self.mating()     # Generate tuples indicating mating pairs\n",
    "            self.crossover()  # Generate two childs for each couple\n",
    "            self.mutate()     # Apply mutation on all individuals (old and new)\n",
    "            self.replace()    # Select which individuals will be kept\n",
    "            \n",
    "            self.gen_count += 1\n",
    "            self.update()\n",
    "            continue\n",
    "        \n",
    "        print(f\"\\nElapsed time: {self.time}\")\n",
    "        self.counter = 0\n",
    "        self.show(self.best, compare=True)\n",
    "        \n",
    "    def update(self):\n",
    "        ''' Updates the fitness of every individual in the population.\n",
    "            Also updates the fittest individual found among all generations.\n",
    "        '''\n",
    "        # Update fittest individual\n",
    "        idx = np.argmin(self.all_fits)\n",
    "        if self.all_fits[idx] < self.best_fit: \n",
    "            self.best     = self.population[idx].copy() # save best inividual\n",
    "            self.best_gen = self.gen_count\n",
    "            self.best_fit = self.all_fits[idx]\n",
    "\n",
    "    def read(self, filepath, res=None):\n",
    "        '''Reads an image in grey scale, resizes it and extends to array.\n",
    "            Params:\n",
    "                filepath - Path to the input image to be read\n",
    "                res - Either tuple with new pixel dimensions or ratio with new scale\n",
    "            Return: \n",
    "                np.array - array with the streched image\n",
    "        '''\n",
    "        img = imread(filepath,0) # load in greyscale\n",
    "        \n",
    "        if res and tuple==type(res):\n",
    "            img = resize(img, res)\n",
    "        elif res and float==type(res):\n",
    "            w = int(img.shape[0]*res)\n",
    "            h = int(img.shape[1]*res)\n",
    "            img = resize(img, (h,w))\n",
    "            \n",
    "        self.shape = img.shape # keep shape\n",
    "        self.goal  = img.reshape(img.size)\n",
    "\n",
    "    def show(self, individual, compare=False):\n",
    "        '''Prints the image represented by an individual.\n",
    "            Params:\n",
    "                individual - Np.array to be reshaped and printed\n",
    "        '''\n",
    "        plt.figure(figsize=(15,15))\n",
    "        \n",
    "        # If compare mode, print goal along side\n",
    "        if compare: \n",
    "            plt.subplot(1,2,2)\n",
    "            goal = self.goal.copy()\n",
    "            goal = goal.reshape(self.shape)\n",
    "            plt.imshow(goal, cmap='gray')\n",
    "            plt.title('Goal')\n",
    "            plt.axis('off')\n",
    "            plt.subplot(1,2,1)\n",
    "            \n",
    "        img = individual.copy()\n",
    "        img = img.reshape(self.shape)\n",
    "        plt.imshow(img, cmap='gray')\n",
    "        plt.title('Individual')\n",
    "        plt.axis('off')\n",
    "        plt.show() \n",
    "    \n",
    "    def write(self, individual, path, folder='Outputs'):\n",
    "        name = g + str(self.gen_count) + '_' + basename(path)\n",
    "        imwrite(join(folder, name), individual)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Tests**\n",
    "\n",
    "Now that a few variations of each technique are implemented, it's possible to test multiple combinations of those methods and other parameters, and compare the final results.\n",
    "\n",
    "As previously mentioned, the test cases are **small**, **monochromatic** images. This is due to the amount of possibilities per individual, the convergence can be very slow if the pixel count is too large."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Generation:  42051 (avg. fit: 616.0). Current Best:  616.000 (gen 32052).....\n",
      "Elapsed time: 1592172768.3295836\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA1MAAAGaCAYAAADqwBnJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAP8ElEQVR4nO3da4jl913H8e93HRtJL+mESGJLEzMxaRJWKC5ivNEnKyhEBBGt9VYvJKXgA7VtIlhtrBdSrbEIohFvZdtIVsREjEpCUSkVYxfzQJBIWRtLrXETT9LdJLvpdn4+OCcwuez07Gd2979n9vV6krPzP+fkM7PJ/Pc9/zOzPcYoAAAATs+eqQcAAACsIjEFAAAQEFMAAAABMQUAABAQUwAAAAExBQAAEBBT7Frd/f7uPrC4fWV3H+vur1jicb/X3e/b5vjo7q87k/sAYLfp7s909/6pd8DZJKY4752JT8ZjjP8aY7xmjPGlJe77zjHGB3by7wOAVdHdb+vuf+7uZ7r7fxe339XdPfU2ON+JKQCAC1R3/1xVfbiqfqOqrqiqy6vqnVX1rVX1qgmnwUoQU6yM7n5Hd3+iu3+zu2fd/Z/d/V1bjl/d3f/Q3Ue7+8GqumzLsa9dvDxvrbt/oLs/9ZLn/pnuvn9x+0+6+1e2HHtPd3++u/+7u3/iJY/7++7+qZdu3PLrD3f3Z7v7C919qLu//Yx+UAAg1N2XVNUvV9W7xhh/PsY4Oub+dYzxQ2OME919SXd/pLuPdPdj3f0L3b1n8fhruvvj3f1kdz/R3R/t7tdP+17BuSWmWDXfVFWP1jyUPlhVf7jlZQgfq6pDi2MfqKofO8Vz/FVVvbm7r93ytrcvHv8i3f2dVfXuqvqOqrq2qk735Yb/UlVvqapLF89/sLu/6jSfAwDOhm+uqouq6r5t7vM7VXVJVW1U1Vur6ker6scXx7qqfr2q3lBVN1TVm6rq/WdpK5yXxBSr5rExxh8svvfpT6vqa6rq8u6+sqq+sareN8Y4Mcb4x5pH08uMMZ6t+YnjB6uqFlF1fVXd/wp3//6q+uMxxr+NMZ6p0zxJjDEOjDGeHGOcHGN8qOYnrTefznMAwFlyWVU9McY4+cIbuvuT3f1Udz/X3W+tqrdV1c8vrlp9pqo+VFU/UlU1xvj0GOPBxXn3SFX9Vs2DCy4YYopV8z8v3FhEUVXVa2r+VbHZInhe8Ng2z/OxWsRUza9K/eWW59vqDVX12SWf82W6+93d/e/d/XR3P1Xzr+5d9uUeBwDnwJNVdVl3r73whjHGt4wxXr84dkVVfWW9+Nz3WFW9saqquy/v7j/r7s919xeq6kA5x3GBEVPsFp+vqvXufvWWt125zf0frKqv7u631DyqXvYSvy3P+6ZtnvOZqrp4y6+veOHG4vuj3lvzq1vri5PT0zV/WQQATO2fqupEVX3PKY4/UVVfrKqrtrztyqr63OL2r1XVqKqvH2O8rqp+uJzjuMCIKXaFMcZjVfWpqrqju1/V3d9WVd+9zf2/WFUHa/7Tiy6teVy9knur6h3dfWN3X1xVv/SS449U1fd298WLv3vqJ7cce21VnayqI1W11t2/WFWvO/33DgDOvDHGU1V1R1X9bnd/X3e/trv3LL7Q+Oqq+lLNz4O/ujh2VVX9bM2vQFXNz3PHqurp7n5jVb3n3L8XMC0xxW7y9pr/gIr/q3n0fOTL3P9jNf+BEge3vl58qzHG31TVb1fVx6vq04t/bnVXVT1fVY/X/Hu4Prrl2N9V1d9W1X/U/GURx+vFLxkEgEmNMT5Y80B6b83PZY9X1e9X1W1V9cmq+umavwrjcFV9oubnzj9aPPyOqvqGmr/q4q+r6i/O5XY4H/QYY+oNAAAAK8eVKQAAgICYAgAACIgpAACAgJgCAAAIrG138ODBgyvx0yn27ds39YSlzWazqScsZX19feoJsGscPnx46glL2b9/v78f5jTcd999K3GOvPHGG6eeAHBKjz766NQTlnLzzTe/4jnSlSkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAwNp2Bzc2Ns7Vjh1ZlZ3AhWk2m009gbPgqquumnrCUq699tqpJwCc0okTJ6aesCOuTAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBgbbuDs9nsXO24YPiYnlnr6+tTT1ja5ubm1BOWsmfPanyNxf9LTO3YsWNTT1jKs88+O/UEgFNalc+lp7Iaf2oCAAA4z4gpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAisTT3gTLj77runnrC0W2+9deoJS1lfX596wlJuu+22qScs7c4775x6wq5y7733Tj2BC9zRo0ennrCUAwcOTD1haatyjrz00kunnrCU22+/feoJS7vrrrumnrCUzc3NqScs5Z577pl6wtJOnjw59YQdcWUKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAILC23cHZbHauduzIQw89NPWEpW1sbEw9YVc5ePDg1BOW5vf+zDp06NDUE5bm9353Onr06NQTlvLAAw9MPWFpe/funXrCrnL//fdPPWFpV1999dQTlrK5uTn1hKU8/PDDU09Y2nXXXTf1hB1xZQoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACYgoAACAgpgAAAAJiCgAAICCmAAAAAmIKAAAgIKYAAAACa1MPOBP2798/9QQmsm/fvqknLO3QoUNTT9hVNjY2pp7ABe748eNTT1iKc+SZd9FFF009YSk33XTT1BOW9sgjj0w9YSmbm5tTT1jK9ddfP/WEpR09enTqCTviyhQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAQEwBAAAExBQAAEBATAEAAATEFAAAQEBMAQAABMQUAABAYG27g7PZ7Fzt2JGNjY2pJyztlltumXoCE9m3b9/UE5jI4cOHp57AWXD8+PGpJyzlhhtumHrC0q655pqpJyxlbW3bPz6dN/bsWZ2vme/du3fqCbvKKv3eHzlyZOoJO7I6H2kAAIDziJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgMDadgdns9m52rEjq7ITgN3jueeem3rCUo4dOzb1hF3n+eefn3oC7Bqr8rn0VFyZAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgMDadgdns9m52rEjq7KzanW2rq+vTz1hKavy8eTC5r/T3enxxx+fegLAyjty5MjUE3bElSkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgICYAgAACIgpAACAQI8xpt4AAACwclyZAgAACIgpAACAgJgCAAAIiCkAAICAmAIAAAiIKQAAgMD/A3VV389DzYeXAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1080x1080 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ga = GeneticAlgorithm(\n",
    "    limit     = 100000,\n",
    "    fit       = squared_error,\n",
    "    conv      = (convergence,(10000,)), \n",
    "    init      = (random,(100,)),\n",
    "    mating    = (ranked,None), \n",
    "    crossover = (k_point_crossover,[3]), \n",
    "    mutate    = (reduced_mut,[0.05,0.025]), # Tip higher MR and lower MA\n",
    "    replace   = (elitism,None)\n",
    "    )\n",
    "\n",
    "ga.test('images/flickr.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Result Analysis and Comparisons**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notes:\n",
    "\n",
    " - Uniform Crossover is extremely slow since it has to iterate through every gene (and there are a lot of genes) and randomize the possibility of gene exchange. Despite being a possible counter measure to premature convergence, it was simply to slow to use it.\n",
    " - Rergarding mutation rate and mutation ammount, in our case it appeared to be better when we would use the a higher probability of mutation with a smaller number of genes to mutate. The more genes we mutate, the slower is the iteration (as we need to roll the dice more times), so a higher change of mutation was a more efficient approach agains a higher number of genes to mutate.\n",
    " - Different sizes of populations did not seem to interfere significantly with the results, as expected a smaller population converges faster but it alsos converges prematurely most times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Conclusions**\n",
    "\n",
    "As mentioned previously, the amount of possibilities makes this a hard problem to solve, especially for higher resolution images."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
